File: LeetCode Problems/Linked List/Reverse Linked List.py
'''
Reverse a singly linked list.

Example:
Input: 1->2->3->4->5->NULL
Output: 5->4->3->2->1->NULL
'''


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseList(self, head): # recursion
        if not head or not head.next: # when the end of list is reached by recursion
            return head 
        
        next = head.next # head=4, next=5
        newHead = self.reverseList(next) # traverse all the way to end of list, newHead=5
        next.next = head # 5.next = 4,  5--->4
        head.next = None # 5-->4-->None
        
        return newHead
    
    def reverseList(self, head): # iteration
        prev = None
        while head:
            next = head.next
            head.next = prev
            prev = head
            head = next
        return prev

File: LeetCode Problems/C++/Reverse Linked List.cpp
/*
Reverse a singly linked list.

Example:
    Input: 1->2->3->4->5->NULL
    Output: 5->4->3->2->1->NULL

Follow up: A linked list can be reversed either iteratively or recursively. Could you implement both?
 */

#include <iostream>
 
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}

    void printList() {
        ListNode* s = this;
        while (s) {
            cout << s->val << endl;
            s = s->next;
        }
    }
};

class Solution {
public:
    ListNode* reverseListRec(ListNode* head) {
        // If reached the end, the cur (last) node becomes the newHead  (line 32)
        if (!head || !head->next) { // head and next are both pointers to ListNode
            return head;
        }

        ListNode* n = head->next; // head -> next is equivalent to *head.next
        ListNode* newHead = reverseListRec(n); // keep recursing until return the last node (line 28), WONT change once set

        // newHead->5, n->4, head->3
        n->next = head; 
        head->next = NULL;

        return newHead;
        

    }
};

int main() {
    ListNode L =ListNode(3, &ListNode(4, &ListNode(5)));
    Solution s = Solution();
    ListNode* reversedL = s.reverseListRec(&L);
    reversedL->printList();
}